server/src/test/java/org/elasticsearch/rest/action/document/RestGetSourceActionTests.java

```diff
@@ -49,6 +74,39 @@ public static void cleanupReferences() {
         listener = null;
     }
 
+    /**
+     * test deprecation is logged if type is used in path
+     */
+    public void testTypeInPath() {
+        RestRequest request = new FakeRestRequest.Builder(xContentRegistry())
+            .withMethod(randomFrom(Arrays.asList(Method.GET, Method.HEAD)))
+            .withPath("/some_index/some_type/id/_source")
+            .build();
+        performRequest(request);
+        assertWarnings(RestGetSourceAction.TYPES_DEPRECATION_MESSAGE);
+    }
+
+    /**
+     * test deprecation is logged if type is used as parameter
+     */
+    public void testTypeParameter() {
+        Map<String, String> params = new HashMap<>();
+        params.put("type", "some_type");
+        RestRequest request = new FakeRestRequest.Builder(xContentRegistry())
+            .withMethod(randomFrom(Arrays.asList(Method.GET, Method.HEAD)))
```

Comment:
Since these tests are so fast + simple, maybe we could just test both methods every time. I don't think this will be much extra code, as you could re-use the same `FakeRestRequest.Builder`.